### PR TITLE
Multiple bug fixes

### PR DESCRIPTION
**Pull request by korydraughn**
This PR contains a commit that breaks compatibility with existing deployments (i.e. log4j version change).

Configuration files between log4j 1 and 2 are not compatible.

#### Update
The statement above is no longer true. The issues around log4j configuration and log output have been resolved.

### CODE REVIEW

File: irods-vfs-impl/src/main/java/org/irods/nfsrods/vfs/IRODSVirtualFileSystem.java [hunk 1]
```diff
@@ -215,14 +215,18 @@ public int access(Inode _inode, int _mode) throws IOException
...
-        log_.debug("access - _inode path  = {}", path);
-        log_.debug("access - _mode        = {}", _mode);
-        log_.debug("access - _mode & ACCESS4_READ    = {}", (_mode & ACCESS4_READ));
-        log_.debug("access - _mode & ACCESS4_LOOKUP  = {}", (_mode & ACCESS4_LOOKUP));
-        log_.debug("access - _mode & ACCESS4_MODIFY  = {}", (_mode & ACCESS4_MODIFY));
-        log_.debug("access - _mode & ACCESS4_EXTEND  = {}", (_mode & ACCESS4_EXTEND));
-        log_.debug("access - _mode & ACCESS4_DELETE  = {}", (_mode & ACCESS4_DELETE));
-        log_.debug("access - _mode & ACCESS4_EXECUTE = {}", (_mode & ACCESS4_EXECUTE));
+        log_.debug("""
+        		   access - _inode path             = {}
```
Comment (trel) [inline irods-vfs-impl/src/main/java/org/irods/nfsrods/vfs/IRODSVirtualFileSystem.java:219]: whitespace?

File: irods-vfs-impl/src/main/java/org/irods/nfsrods/vfs/IRODSVirtualFileSystem.java [hunk 2]
```diff
@@ -683,30 +707,40 @@ public Access checkAcl(Subject _subject, Inode _inode, int _accessMask) throws C
...
-        log_.debug("checkAcl - _accessMask & ACE4_DELETE_CHILD      = {}", _accessMask & ACE4_DELETE_CHILD);
-        log_.debug("checkAcl - _accessMask & ACE4_READ_ATTRIBUTES   = {}", _accessMask & ACE4_READ_ATTRIBUTES);
-        log_.debug("checkAcl - _accessMask & ACE4_WRITE_ATTRIBUTES  = {}", _accessMask & ACE4_WRITE_ATTRIBUTES);
-        log_.debug("checkAcl - _accessMask & ACE4_DELETE            = {}", _accessMask & ACE4_DELETE);
-        log_.debug("checkAcl - _accessMask & ACE4_READ_ACL          = {}", _accessMask & ACE4_READ_ACL);
-        log_.debug("checkAcl - _accessMask & ACE4_WRITE_ACL         = {}", _accessMask & ACE4_WRITE_ACL);
-        log_.debug("checkAcl - _accessMask & ACE4_WRITE_OWNER       = {}", _accessMask & ACE4_WRITE_OWNER);
-        log_.debug("checkAcl - _accessMask & ACE4_SYNCHRONIZE       = {}", _accessMask & ACE4_SYNCHRONIZE);
+        log_.debug("""
+        		   checkAcl - _accessMask & ACE4_READ_DATA         = {}
```
Comment (trel) [inline irods-vfs-impl/src/main/java/org/irods/nfsrods/vfs/IRODSVirtualFileSystem.java:721]: whitespace?

File: irods-vfs-impl/src/main/java/org/irods/nfsrods/vfs/IRODSVirtualFileSystem.java [hunk 3]
```diff
@@ -683,30 +707,40 @@ public Access checkAcl(Subject _subject, Inode _inode, int _accessMask) throws C
         }
         
         // @formatter:off
-        log_.debug("checkAcl - _subject uid         = {}", Subjects.getUid(_subject));
-        log_.debug("checkAcl - _subject primary gid = {}", Subjects.getPrimaryGid(_subject));
-        log_.debug("checkAcl - _inode path          = {}", path);
-        log_.debug("checkAcl - _accessMask          = {}", _accessMask);
-        log_.debug("checkAcl - username             = {}", userName);
+        log_.debug("""
+        		checkAcl - _subject uid         = {}
```
Comment (trel) [inline irods-vfs-impl/src/main/java/org/irods/nfsrods/vfs/IRODSVirtualFileSystem.java:711]: whitespace?

File: irods-vfs-impl/src/main/java/org/irods/nfsrods/vfs/IRODSVirtualFileSystem.java [hunk 4]
```diff
@@ -380,23 +391,30 @@ private UserFilePermission toUserFilePermission(nfsace4 _ace) throws JargonExcep
...
-        log_.debug("toUserFilePermission - _accessMask & ACE4_DELETE_CHILD      = {}", accessMask & ACE4_DELETE_CHILD);
-        log_.debug("toUserFilePermission - _accessMask & ACE4_READ_ATTRIBUTES   = {}", accessMask & ACE4_READ_ATTRIBUTES);
-        log_.debug("toUserFilePermission - _accessMask & ACE4_WRITE_ATTRIBUTES  = {}", accessMask & ACE4_WRITE_ATTRIBUTES);
-        log_.debug("toUserFilePermission - _accessMask & ACE4_DELETE            = {}", accessMask & ACE4_DELETE);
-        log_.debug("toUserFilePermission - _accessMask & ACE4_READ_ACL          = {}", accessMask & ACE4_READ_ACL);
-        log_.debug("toUserFilePermission - _accessMask & ACE4_WRITE_ACL         = {}", accessMask & ACE4_WRITE_ACL);
-        log_.debug("toUserFilePermission - _accessMask & ACE4_WRITE_OWNER       = {}", accessMask & ACE4_WRITE_OWNER);
-        log_.debug("toUserFilePermission - _accessMask & ACE4_SYNCHRONIZE       = {}", accessMask & ACE4_SYNCHRONIZE);
+        log_.debug("""
+        		   toUserFilePermission - _accessMask & ACE4_READ_DATA         = {}
```
Comment (trel) [inline irods-vfs-impl/src/main/java/org/irods/nfsrods/vfs/IRODSVirtualFileSystem.java:395]: whitespace?

File: irods-vfs-impl/src/main/java/org/irods/nfsrods/vfs/IRODSVirtualFileSystem.java [hunk 5]
```diff
@@ -358,10 +367,12 @@ private UserFilePermission toUserFilePermission(nfsace4 _ace) throws JargonExcep
     {
         String who = _ace.who.toString().split("[@]")[0];
 
-        log_.debug("toUserFilePermission - ace who         = {}", who);
-        log_.debug("toUserFilePermission - ace type        = {}", _ace.type.value.value);
-        log_.debug("toUserFilePermission - ace flag        = {}", _ace.flag.value.value);
-        log_.debug("toUserFilePermission - ace access mask = {}", _ace.access_mask.value.value);
+        log_.debug("""
+        		   toUserFilePermission - ace who         = {}
```
Comment (trel) [inline irods-vfs-impl/src/main/java/org/irods/nfsrods/vfs/IRODSVirtualFileSystem.java:371]: whitespace?

File: irods-vfs-impl/src/main/java/org/irods/nfsrods/vfs/IRODSVirtualFileSystem.java [hunk 6]
```diff
@@ -246,13 +250,18 @@ public Inode create(Inode _parent, Type _type, String _name, Subject _subject, i
...
 
-        log_.debug("create - _parent      = {}", parentPath);
-        log_.debug("create - _type        = {}", _type);
-        log_.debug("create - _name        = {}", _name);
-        log_.debug("create - _subject     = {}", _subject);
-        log_.debug("create - _subject uid = {}", Subjects.getUid(_subject));
-        log_.debug("create - _subject gid = {}", Subjects.getPrimaryGid(_subject));
-        log_.debug("create - _mode        = {}", Stat.modeToString(_mode));
+        log_.debug("""
+        		   create - _parent      = {}
```
Comment (trel) [inline irods-vfs-impl/src/main/java/org/irods/nfsrods/vfs/IRODSVirtualFileSystem.java:254]: whitespace?

File: irods-vfs-impl/src/main/java/org/irods/nfsrods/vfs/IRODSVirtualFileSystem.java [hunk 7]
```diff
@@ -991,9 +1027,11 @@ public Inode lookup(Inode _parent, String _path) throws IOException
         Path parentPath = getPath(toInodeNumber(_parent));
         Path targetPath = parentPath.resolve(_path);
 
-        log_.debug("lookup - _path   = {}", _path);
-        log_.debug("lookup - _parent = {}", parentPath);
-        log_.debug("lookup - Looking up [{}] ...", targetPath);
+        log_.debug("""
+        		lookup - _path   = {}
```
Comment (trel) [inline irods-vfs-impl/src/main/java/org/irods/nfsrods/vfs/IRODSVirtualFileSystem.java:1031]: whitespace?

File: irods-vfs-impl/src/main/java/org/irods/nfsrods/vfs/IRODSVirtualFileSystem.java [hunk 8]
```diff
@@ -1091,10 +1129,12 @@ public boolean move(Inode _srcParentInode, String _srcFilename, Inode _dstParent
         Path srcParentPath = getPath(toInodeNumber(_srcParentInode));
         Path dstParentPath = getPath(toInodeNumber(_dstParentInode));
 
-        log_.debug("move - _inode path (src) = {}", srcParentPath);
-        log_.debug("move - _inode path (dst) = {}", dstParentPath);
-        log_.debug("move - _srcName          = {}", _srcFilename);
-        log_.debug("move - _dstName          = {}", _dstFilename);
+        log_.debug("""
+        		   move - _inode path (src) = {}
```
Comment (trel) [inline irods-vfs-impl/src/main/java/org/irods/nfsrods/vfs/IRODSVirtualFileSystem.java:1133]: whitespace?

File: irods-vfs-impl/src/main/java/org/irods/nfsrods/vfs/IRODSVirtualFileSystem.java [hunk 9]
```diff
@@ -1185,10 +1227,12 @@ public int read(Inode _inode, byte[] _data, long _offset, int _count) throws IOE
         {
             Path path = getPath(toInodeNumber(_inode));
 
-            log_.debug("read - _inode path  = {}", path);
-            log_.debug("read - _data.length = {}", _data.length);
-            log_.debug("read - _offset      = {}", _offset);
-            log_.debug("read - _count       = {}", _count);
+            log_.debug("""
+            		   read - _inode path  = {}
```
Comment (trel) [inline irods-vfs-impl/src/main/java/org/irods/nfsrods/vfs/IRODSVirtualFileSystem.java:1231]: whitespace?

File: irods-vfs-impl/src/main/java/org/irods/nfsrods/vfs/IRODSVirtualFileSystem.java [hunk 10]
```diff
@@ -1263,9 +1309,11 @@ public void remove(Inode _parent, String _path) throws IOException
...
     public void setattr(Inode _inode, Stat _stat) throws IOException
     {
-        log_.debug("vfs::setattr");
-        log_.debug("setattr - _inode = {}", getPath(toInodeNumber(_inode)));
-        log_.debug("setattr - _stat  = {}", _stat);
+        log_.debug("""
+        		   vfs::setattr
+                   setattr - _inode = {}
+                   setattr - _stat  = {}""",
+        		   getPath(toInodeNumber(_inode)), _stat);
```
Comment (trel) [inline irods-vfs-impl/src/main/java/org/irods/nfsrods/vfs/IRODSVirtualFileSystem.java:1316]: whitespace?

File: irods-vfs-impl/src/main/java/org/irods/nfsrods/vfs/IRODSVirtualFileSystem.java [hunk 11]
```diff
@@ -1295,10 +1343,12 @@ public WriteResult write(Inode _inode, byte[] _data, long _offset, int _count, S
         {
             Path path = getPath(toInodeNumber(_inode));
 
-            log_.debug("write - _inode path  = {}", path);
-            log_.debug("write - _data.length = {}", _data.length);
-            log_.debug("write - _offset      = {}", _offset);
-            log_.debug("write - _count       = {}", _count);
+            log_.debug("""
+            		   write - _inode path  = {}", path);
```
Comment (trel) [inline irods-vfs-impl/src/main/java/org/irods/nfsrods/vfs/IRODSVirtualFileSystem.java:1347]: whitespace?

File: irods-vfs-impl/src/main/java/org/irods/nfsrods/vfs/IRODSVirtualFileSystem.java [hunk 12]
```diff
@@ -1404,10 +1456,12 @@ private Stat statPath(Path _path, long _inodeNumber) throws IOException
             stat.setFileid((int) _inodeNumber);
             stat.setGeneration(objStat.getModifiedAt().getTime());
 
-            log_.debug("statPath - User ID           = {}", userId);
-            log_.debug("statPath - Group ID          = {}", groupId);
-            log_.debug("statPath - Permissions       = {}", Stat.modeToString(stat.getMode()));
-            log_.debug("statPath - Stat              = {}", stat);
+            log_.debug("""
+            		   statPath - User ID           = {}
```
Comment (trel) [inline irods-vfs-impl/src/main/java/org/irods/nfsrods/vfs/IRODSVirtualFileSystem.java:1460]: whitespace?

File: irods-vfs-impl/src/main/java/org/irods/nfsrods/vfs/IRODSVirtualFileSystem.java [hunk 13]
```diff
@@ -1426,9 +1480,11 @@ private Stat statPath(Path _path,
...
         throws IOException
     {
-        log_.debug("statPath - _inodeNumber          = {}", _inodeNumber);
-        log_.debug("statPath - _path                 = {}", _path);
-        log_.debug("statPath - iRODS permissions     = {}", _entry.getUserFilePermission());
+        log_.debug("""
+        		   statPath - _inodeNumber          = {}
+                   statPath - _path                 = {}
+                   statPath - iRODS permissions     = {}""",
+        		   _inodeNumber, _path, _entry.getUserFilePermission());
```
Comment (trel) [inline irods-vfs-impl/src/main/java/org/irods/nfsrods/vfs/IRODSVirtualFileSystem.java:1487]: whitespace?

File: irods-vfs-impl/src/main/java/org/irods/nfsrods/vfs/IRODSVirtualFileSystem.java [hunk 14]
```diff
@@ -1479,10 +1535,12 @@ private Stat statPath(Path _path,
             stat.setFileid((int) _inodeNumber);
             stat.setGeneration(_entry.getModifiedAt().getTime());
 
-            log_.debug("statPath - User ID           = {}", userId);
-            log_.debug("statPath - Group ID          = {}", groupId);
-            log_.debug("statPath - Permissions       = {}", Stat.modeToString(stat.getMode()));
-            log_.debug("statPath - Stat              = {}", stat);
+            log_.debug("""
+            		   statPath - User ID           = {}
```
Comment (trel) [inline irods-vfs-impl/src/main/java/org/irods/nfsrods/vfs/IRODSVirtualFileSystem.java:1539]: whitespace?

**Review by trel — COMMENTED** [review]
seems... unlikely you goofed 14 times... but... there are other places where things line up... so i'm not sure what's happening in this commit

File: irods-vfs-impl/src/main/java/org/irods/nfsrods/vfs/IRODSVirtualFileSystem.java (hunk 1 above)
Comment (korydraughn) [reply irods-vfs-impl/src/main/java/org/irods/nfsrods/vfs/IRODSVirtualFileSystem.java:219]: vim vs eclipse formatting. i'll fix these before we merge this PR.

**Comment by korydraughn** [comment]
Items remaining for this PR:
- [ ] Fix issues with truncate command.
- [ ] Investigate whether other cache entries need to be evicted upon data movement.
- [ ] Review README (this can be done in another PR).
- [ ] Add BATs tests.
- [ ] Cleanup whitespace formatting in Java source files.

Am I missing anything?

**Comment by trel** [comment]
yeah, remove README from this list - different PR.

otherwise, seems good.

File: start.sh [hunk 15]
```diff
@@ -15,6 +15,7 @@ if [ "$1" != "sha" ]; then
     fi
 
     # Start services.
+    [ ! -d /run/sendsigs.omit.d ] && mkdir /run/sendsigs.omit.d
```
Comment (trel) [inline start.sh:18]: how is this different than just `mkdir -p`?

Comment (korydraughn) [reply start.sh:18]: i forgot about the -p option. this line still needs more investigation. it keeps an unexpected line from appearing in the log file, but the server is acting weird now.

**Comment by korydraughn** [comment]
I need to revisit the docker files. The design is a little janky in regards to building and running. The repo should likely separate the build step from the run step (similar to how Metalnx handles it).

Changes:
- Provide instructions or a convenience script for building NFSRODS against the local repo.
  - This would use a maven docker image.
- Update the Dockerfile to copy the previously built JAR file into the docker image.

Essentially, the Dockerfile should not have to clone the repo at all.

**Comment by korydraughn** [comment]
Possible things to handle in a separate PR:
- truncate / jargon
- performance test numbers
- ubuntu 16 docker container issues

File: irods-vfs-impl/pom.xml [hunk 16]
```diff
@@ -204,10 +248,53 @@
...
+
+						<artifactItem>
+							<groupId>${project.groupId}</groupId>
+							<artifactId>${project.artifactId}</artifactId>
+							<version>${project.version}</version>
+							<classifier>jar-with-dependencies</classifier>
+							<type>${project.packaging}</type>
+							<overWrite>true</overWrite>
+							<outputDirectory>${project.basedir}/..</outputDirectory>
+<!-- 							<destFileName>nfsrods</destFileName> -->
```
[the remark this answers is not in the document]
Comment (korydraughn) [reply irods-vfs-impl/pom.xml:287]: removing the version number from the JAR file would decrease the number of files needing updates due when releasing  future versions of nfsrods.

File: testing/behavior_tests.bats [hunk 17]
```diff
@@ -3,24 +3,23 @@
...
+#   4. bats behaviour_tests.bat
 #
 
 setup() {
     cd /mnt/nfsrods/home/rods
 }
 
 @test "test list contents of collection" {
-    ls /mnt/nfsrods/home/rods
+    /bin/ls /mnt/nfsrods/home/rods
```
Comment (trel) [inline testing/behavior_tests.bats:22]: not essential.... consider defining an `$LS_EXECUTABLE` and using that everywhere in this file

File: irods-vfs-impl/pom.xml (hunk 16 above)
[the remark this answers is not in the document]
Comment (korydraughn) [reply irods-vfs-impl/pom.xml:287]: after some discussion, the plan is to remove version numbers from the copied nfsrods JAR file.

File: conformance_testing/test_nfsrods.bats [hunk 18]
```diff
@@ -1,58 +1,199 @@
...
+SANDBOX=nfsrods_bats_testing_sandbox
+
+setup() {
+    mkdir -p $SANDBOX
+    cd $SANDBOX
+}
+
+teardown() {
+    cd ..
+    rm -rf $SANDBOX
```
Comment (trel) [inline conformance_testing/test_nfsrods.bats:34]: please use `${SANDBOX}` for these three

File: conformance_testing/test_nfsrods.bats [hunk 19]
```diff
@@ -1,58 +1,199 @@
...
-@test "create, list, and remove file" {
-    FILENAME=removeme
+@test "create, list, rename, move, copy, and remove file" {
+    local FILENAME=removeme
 
     run echo ${FILENAME} > ${FILENAME}
     [ $status -eq 0 ]
 
-    run ls ${FILENAME}
+    run $LS_EXECUTABLE ${FILENAME}
```
Comment (trel) [inline conformance_testing/test_nfsrods.bats:43]: `${LS_EXECUTABLE}`

File: conformance_testing/test_nfsrods.bats [hunk 20]
```diff
@@ -1,58 +1,199 @@
...
+    run cp ${NEW_FILENAME} col.d/${FILENAME}
+    [ $status -eq 0 ]
+
+    run rm col.d/${FILENAME}
+    [ $status -eq 0 ]
+
+    run rm ${NEW_FILENAME}
+    [ $status -eq 0 ]
+
+    run 2 $LS_EXECUTABLE ${NEW_FILENAME}
```
Comment (trel) [inline conformance_testing/test_nfsrods.bats:61]: `${LS_EXECUTABLE}`

File: conformance_testing/test_nfsrods.bats [hunk 21]
```diff
@@ -1,58 +1,199 @@
...
+    #[ $status -eq 0 ]
 
-    run mkdir -p ${SANDBOX}/c{001..125}
+    #run mkdir -p ${SANDBOX}/c{001..125}
+    run mkdir -p c{001..125}
     [ $status -eq 0 ]
 
-    run ls ${SANDBOX}
+    #run $LS_EXECUTABLE ${SANDBOX}
+    run $LS_EXECUTABLE
```
Comment (trel) [inline conformance_testing/test_nfsrods.bats:80]: `${LS_EXECUTABLE}`

File: conformance_testing/test_nfsrods.bats [hunk 22]
```diff
@@ -1,58 +1,199 @@
...
+
+    #run parallel rm ::: ${SANDBOX}/foo{00001..10000}
+    run parallel rm ::: foo{0001..6000}
+    #rm -rf ${SANDBOX}
+}
+
+@test "create and remove directory" {
+    local COLLECTION='col.d'
+    mkdir $COLLECTION
+    rmdir $COLLECTION
```
Comment (trel) [inline conformance_testing/test_nfsrods.bats:108]: `${COLLECTION}`

File: conformance_testing/test_nfsrods.bats [hunk 23]
```diff
@@ -1,58 +1,199 @@
...
+}
+
+@test "write and read non-empty file" {
+    local FILENAME=bar
+
+    echo -n 'Hello, NFSRODS!' > $FILENAME
+
+    run cat $FILENAME
```
Comment (trel) [inline conformance_testing/test_nfsrods.bats:116]: `${FILENAME}`

**Review by trel — COMMENTED** [review]
please add `{}` around all the bash variables.   i got tired ... there are more :P

File: irods-vfs-impl/src/main/java/org/irods/nfsrods/vfs/IRODSVirtualFileSystem.java [hunk 24]
```diff
@@ -111,6 +115,15 @@
     private final ReadWriteAclWhitelist readWriteAclWhitelist_;
     private final boolean allowOverwriteOfExistingFiles_;
     private final boolean usingOracleDB_;
+    
+    // Maintains a set of locks for parallel writes for each (user, path)
+    // tuple. The locks are used to synchronize calls to open and close across
+    // parallel writes to the same replica. This is unique to NFSRODS because
+    // it cannot know how the client app (e.g. dd) will attempt to write data.
+    // Ultimately, the locks are used to avoid issues around replica access
+    // tokens (e.g. a call to open fails due to not having the required to
```
Comment (trel) [inline irods-vfs-impl/src/main/java/org/irods/nfsrods/vfs/IRODSVirtualFileSystem.java:124]: remove an extra 'to' at the end.

Comment (korydraughn) [reply irods-vfs-impl/src/main/java/org/irods/nfsrods/vfs/IRODSVirtualFileSystem.java]: removed entire comment due to pending jargon bug fixes.

see https://github.com/DICE-UNC/jargon/pull/397

**Comment by korydraughn** [comment]
passed all tests.

added #'s

**Pull request merged**